specification.md

```diff
@@ -2852,6 +2872,163 @@ CustomMessage {
 }
 ```
 
+### AvailableComponents Message
+
+Status: [Development]
+
+Each Agent may be composed of multiple sub-components. These components may have
+their own associated metadata. The AvailableComponents message allows the Agent
+to inform the OpAMP server about which components the Agent contains.
+
+The AvailableComponents message has the following structure:
+
+```protobuf
+message AvailableComponents {
+    map<string, ComponentDetails> components = 1;
+    bytes hash = 2;
+ }
+```
+
+#### AvailableComponents.components
+
+The components field contains a map of a unique ID
+to a [ComponentsDetails](#componentdetails-message) message.
+This field may be omitted from the message if the OpAMP server has not explicitly
+requested it by setting the ReportAvailableComponents flag in the preceding
+ServerToAgent message.
+
+##### Examples
+
+###### OpenTelemetry Collector
+
+Here is an example of how ComponentDetails could hold information regarding the included
+components for a custom build of [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector):
+
+```jsonc
+{
+  "receivers": {
+    "sub_component_map": {
+      "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver@v0.107.0": {
+        "metadata": {
+          "type": "hostmetrics",
+        }
+      }
+    }
+  },
+  "processors": {
+    "sub_component_map": {
+      "go.opentelemetry.io/collector/processor/batchprocessor@v0.107.0": {
+        "metadata": {
+          "type": "batch",
+        }
+      },
+      "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor@v0.107.0": {
+        "metadata": {
+          "type": "transform",
+        }
+      },
+    }
+  },
+  "exporters": {
+    "sub_component_map": {
+      "go.opentelemetry.io/collector/exporter/nopexporter@v0.107.0": {
+        "metadata": {
+          "type": "nop",
+        }
+      }
+    }
+  }
+  // ... Component list continues for extensions and collectors ...
+}
+```
+
+###### Fluent Bit
+
+Here's an example of how Fluent Bit could report what components it has available.
+
+```jsonc
+{
+  "input": {
+    "sub_component_map": {
+      "tail": {}
+    }
+  },
+  "parser": {
+    "sub_component_map": {
+      "json": {}
+    }
+  },
+  "filter": {
+    "sub_component_map": {
+      "lua": {},
+      "modify": {}
+    }
+  },
+  "output": {
+    "sub_component_map": {
+      "null": {},
+      "file": {},
+    }
+  }
+}
+```
+
+#### AvailableComponents.hash
+
+The agent-calculated hash of the components field. The agent MUST include this hash
```

Comment:
At the end of the day this is an opaque value from the server's point of view, but it would be nice to have some guidance on what this may look like. We should probably at least call out some basics of what properties we expect from the hash ("SHOULD be unique to a particular build of an agent, not to an instance") and that its format is determined by the agent.

Reply:
I added that the agent hash should be the same if the components are the same. 

I want to avoid saying it should be unique for a build of the agent, since I'd like this to also support plugin-based agents where a particular build might be the same but the plugins on the filesystem might differ.